services/jupyter/src/app/model/train_model.py

```diff
@@ -0,0 +1,324 @@
+        model_ckpt, num_labels=_num_labels
+    )
+
+    # model.from_pretrained(model_ckpt, num_labels=_num_labels)
+
+    return model
+
+
+# %%
+from sklearn.metrics import accuracy_score, f1_score
+
+
+def compute_metrics(pred):
+    labels = pred.label_ids
+    preds = pred.predictions.argmax(-1)
+    f1 = f1_score(labels, preds, average="weighted")
+    acc = accuracy_score(labels, preds)
+    return {"accuracy": acc, "f1": f1}
+
+
+from transformers import Trainer, TrainingArguments
+
+
+def test_model(model, test_data, tokenizer):
+
+    trainer = Trainer(
+        model=model,
+        tokenizer=tokenizer,
+        compute_metrics=compute_metrics,
+    )
+
+    model.eval()
+
+    trainer.predict(test_data)
+
+    return trainer
+
+
+def train_model(
+    model,
+    wandb_name: str,
+    model_ckpt: str,
+    epochs: int = 1,
+    batch_size: int = 64,
+):
+
+    train_dataset, val_dataset, _ = load_data()
+
+    tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
+
+    def tokenize(batch):
+        return tokenizer(batch["text"], padding=True, truncation=True)
+
+    train_encoded = train_dataset.map(tokenize, batched=True, batch_size=None)
+    val_encoded = val_dataset.map(tokenize, batched=True, batch_size=None)
+
+    logging_steps = len(train_encoded) // batch_size
+    model_name = f"{model_ckpt}-finetuned-news"
+
+    training_args = TrainingArguments(
+        output_dir=model_name,
+        num_train_epochs=epochs,
+        learning_rate=2e-5,
+        per_device_train_batch_size=batch_size,
+        per_device_eval_batch_size=batch_size,
+        weight_decay=0.01,
+        evaluation_strategy="epoch",
+        # optim="adamw_torch",
+        disable_tqdm=False,
+        logging_steps=logging_steps,
+        push_to_hub=True,
+        report_to="wandb",
+        run_name=wandb_name,  # "distill-bert-base-config",
+        log_level="error",
+        hub_token=settings.HF_TOKEN,
+    )
+
+    trainer = Trainer(
+        model=model,
+        args=training_args,
+        compute_metrics=compute_metrics,
+        train_dataset=train_encoded,
+        eval_dataset=val_encoded,
+        tokenizer=tokenizer,
+    )
+
+    trainer.train()
+
+    return trainer
+
+
+class WandbModelArtifact(BaseModel):
+
+    entity: str
+    project: str
+    artifact_name: str
+    tag: str
+
+    def __str__(self):
+        return f"{self.entity}/{self.project}/{self.artifact_name}:{self.tag}"
+
+
+NUM_LABELS = 4
+
+
+def upload_to_s3_bucket(
+    torchscript_dir: str = "model.pt",
+    bucket_name: str = "fsdl-model-test",
+    s3_filename: str = "model.pt",
+):
+    s3 = boto3.client(
+        "s3",
+        settings.AWS_REGION,
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+    )
+
+    # upload file from local directory to s3 bucket
+    with open(torchscript_dir, "rb") as f:
+        s3.upload_fileobj(f, bucket_name, s3_filename)
```

Comment:
at some point we will need to migrate the CLI to the jupyter container to not rely on the docker host. I think it can stay since it's not being called.